source/extensions/filters/http/rate_limit_quota/filter.cc

```diff
@@ -2,28 +2,39 @@
 
 #include <memory>
 
+#include "source/extensions/filters/http/rate_limit_quota/matcher.h"
+
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace RateLimitQuota {
 
-void RateLimitQuotaFilter::setDecoderFilterCallbacks(
-    Http::StreamDecoderFilterCallbacks& callbacks) {
-  callbacks_ = &callbacks;
-}
-
-// TODO(tyxia) Mostly are example/boilerplate code, polish implementation here.
 Http::FilterHeadersStatus RateLimitQuotaFilter::decodeHeaders(Http::RequestHeaderMap& headers,
                                                               bool) {
-  // TODO(tyxia) Create the rate limit gRPC client and start the stream on the first request.
-  absl::StatusOr<BucketId> match_result = requestMatching(headers);
-
-  // Request is not matched by any matchers. In this case, requests are ALLOWED by default (i.e.,
-  // fail-open) and will not be reported to RLQS server.
+  // First, perform the request matching.
+  absl::StatusOr<Matcher::ActionPtr> match_result = requestMatching(headers);
   if (!match_result.ok()) {
+    // When the request is not matched by any matchers, it is ALLOWED by default (i.e., fail-open)
```

Comment:
We will definitely want to add a counter here to track fail open requests.

Reply:
Very good point. Is it ok that I add the stats in a separate PR? Because it will involve set up stats infra e.g., Stats::Scope, COUNTER enum. Also, I want to scan through the whole filter and add more stats than just this place (if needed).

I added a TODO here for stats. Please let me know WDYT.